storage/core/src/main/java/com/hortonworks/streamline/storage/CacheBackedStorageManager.java

```diff
@@ -52,32 +52,51 @@ public void init(Map<String, Object> properties) {
     @Override
     public void add(Storable storable) throws StorageException {
         writer.add(storable);
-        cache.put(storable.getStorableKey(), storable);
+        if (storable.isCacheable()) {
+            cache.put(storable.getStorableKey(), storable);
+        }
     }
 
     @Override
     public <T extends Storable> T remove(StorableKey key) throws StorageException {
-        writer.remove(key);
-        final T oldVal = (T) cache.get(key);
-        cache.remove(key);
-        return oldVal;
+        Storable storable = (Storable) writer.remove(key);
+        if (storable != null && storable.isCacheable()) {
+            storable = cache.get(key);
+            cache.remove(key);
+        }
+        return (T) storable;
     }
 
     @Override
     public void addOrUpdate(Storable storable) throws StorageException {
         writer.addOrUpdate(storable);
-        cache.put(storable.getStorableKey(), storable);
+        if (storable.isCacheable()) {
+            cache.put(storable.getStorableKey(), storable);
+        }
+    }
+
+    @Override
+    public void update(Storable storable) {
+        writer.update(storable);
+        if (storable.isCacheable()) {
+            cache.put(storable.getStorableKey(), storable);
+        }
     }
 
     @Override
     public <T extends Storable> T get(StorableKey key) throws StorageException {
-        return (T) cache.get(key);
+        Storable storable = cache.get(key);
+        // we could load the entries via a LoadingCache so need to explicitly remove it
+        if (storable != null && !storable.isCacheable()) {
```

Comment:
I think we assumed a loading cache here so we don't explicitly retrieve the value from db and put it into the cache. With the existing loading cache implementation its not straightforward to instruct loading cache to load but not cache and still return the value so I would like to defer this to later.